libraries/GCS_MAVLink/MissionItemProtocol.cpp

```diff
@@ -25,9 +25,28 @@ void MissionItemProtocol::init_send_requests(GCS_MAVLINK &_link,
 void MissionItemProtocol::handle_mission_clear_all(const GCS_MAVLINK &_link,
                                                    const mavlink_message_t &msg)
 {
-    bool success = true;
-    success = success && !receiving;
-    success = success && clear_all_items();
+    if (receiving) {
+        // someone is already uploading a mission.  If we are
+        // receiving from someone then we will allow them to clear -
+        // otherwise we deny.
+        if (msg.sysid != dest_sysid || msg.compid != dest_compid) {
+            send_mission_ack(_link, msg, MAV_MISSION_DENIED);
+            return;
+        }
+        free_upload_resources();
+    }
+
+    bool success = clear_all_items();
+
+    if(receiving)
+    {
+        // close down previous upload connection
+        send_mission_ack(_link, msg, success ? MAV_MISSION_ACCEPTED : MAV_MISSION_ERROR);
+        receiving = false;
+        link = nullptr;
```

Comment:
Why is this state not being reset in the same spot you're doing the `free_upload_resources`?

It would seem to simplify things if it were that way, so I can only imagine there's a reason you've done it this way.  If there is it deserves a comment.